### PR TITLE
add conv2d module

### PR DESCRIPTION
**Pull request by mosout**

### CODE REVIEW

File: oneflow/python/test/modules/test_conv.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                [-0.03205633, -11.674841, -10.99459],
+            ]
+        ).reshape((1, 3, 3, 3))
+        conv = flow.nn.Conv2d(1, 3, (3, 3))
+        x = flow.Tensor(input_arr)
+        flow.nn.init.constant_(conv.weight, 2.3)
+        flow.nn.init.constant_(conv.bias, 2.3)
+        of_out = conv(x)
+        test_case.assertTrue(
+            np.allclose(of_out.numpy(), output_arr, rtol=1e-2, atol=1e-5)
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:86]: rtol 这么大吗

File: oneflow/python/test/modules/test_conv.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                [-7.1190453, -11.122115, -4.302488],
+                [-2.3320563, -13.974841, -13.29459],
+                [1.4438007, -6.4144573, -0.66943216],
+                [-7.1190453, -11.122115, -4.302488],
+                [-2.3320563, -13.974841, -13.29459],
+            ]
+        ).reshape((1, 3, 3, 3))
+        conv = flow.nn.Conv2d(1, 3, (3, 3), bias=False)
+        x = flow.Tensor(input_arr)
+        flow.nn.init.constant_(conv.weight, 2.3)
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:51]: 这样的测试用例测试力度不够，很多 bug 测不出来，把 weight 和 bias 也用无规律数据代替吧

File: oneflow/python/test/modules/test_conv.py (hunk 1 above)
Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:86]: 这个借鉴的onnx里的范围

Comment (daquexian) [reply oneflow/python/test/modules/test_conv.py:86]: 设个小一些的试试看能不能通过吧，这个 rtol （百分之一）感觉过大了

Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:86]: 好的

File: oneflow/python/test/modules/test_conv.py (hunk 2 above)
Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:51]: 权重和bias用随机数 感觉不是很好对比

Comment (BBuf) [reply oneflow/python/test/modules/test_conv.py:51]: 可以提前生成随机值，然后用torch把结果算出来进行对比。

Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:51]: 好的

File: oneflow/python/test/modules/test_conv.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import unittest
+import numpy as np
+import oneflow as flow
+
+
+@unittest.skipIf(
+    not flow.unittest.env.eager_execution_enabled(),
+    ".numpy() doesn't work in lazy mode",
+)
+class TestConv2d(flow.unittest.TestCase):
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:25]: 再加几个测试用例测 corner case，例如非正方形的 kernel，h 和 w 方向不一样的 stride 和 padding。
还有测一下 dilation 和 group

File: oneflow/python/nn/modules/conv.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        padding_mode: str = "zeros",  # TODO: refine this type
+    ):
+        super().__init__()
+
+        assert padding_mode == "zeros"
+        kernel_size = _pair(kernel_size)
+        stride = _pair(stride)
+        padding = _pair(padding)
+        dilation = _pair(dilation)
+        self.weight = flow.nn.Parameter(
```
Comment (Ldpe2G) [inline oneflow/python/nn/modules/conv.py:57]: weight 和 bias 都没有初始化，https://pytorch.org/docs/stable/_modules/torch/nn/modules/conv.html#Conv2d，可以参考pytorch的实现，加上 `reset_parameters` 函数，参考 `class _ConvNd(Module):` 这个类

File: oneflow/python/test/modules/test_conv.py (hunk 3 above)
Comment (Ldpe2G) [reply oneflow/python/test/modules/test_conv.py:25]: 或者写一个 naive 的 卷积实现，直接和这个实现对比结果，这样子就不用写死 输入输出了

Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:25]: 好的

File: oneflow/python/nn/modules/conv.py (hunk 4 above)
Comment (mosout) [reply oneflow/python/nn/modules/conv.py:57]: 好的

File: oneflow/python/test/modules/test_conv.py [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+    def test_conv2d(test_case):
+        conv = flow.nn.Conv2d(1, 3, (3, 3), bias=False)
+        _test_conv2d(
+            test_case, conv, test_conv2d_data, test_conv2d_output, test_conv2d_weight
+        )
+
+    def test_conv2d_with_bias(test_case):
+        conv = flow.nn.Conv2d(1, 3, (3, 3))
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:724]: 显式传 bias=True

File: oneflow/python/test/modules/test_conv.py [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+        _test_conv2d(
+            test_case,
+            conv,
+            test_conv2d_group_data,
+            test_conv2d_group_output,
+            test_conv2d_group_weight,
+        )
+
+    def test_conv2d_padding(test_case):
+        conv = flow.nn.Conv2d(1, 1, (3, 3), padding=(1, 1))
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:745]: 测个 (1, 2) 的 padding

File: oneflow/python/test/modules/test_conv.py [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+        _test_conv2d(
+            test_case,
+            conv,
+            test_conv2d_padding_data,
+            test_conv2d_padding_output,
+            test_conv2d_padding_weight,
+        )
+
+    def test_conv2d_stride(test_case):
+        conv = flow.nn.Conv2d(1, 1, (3, 3), padding=(1, 1), stride=(2, 2))
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:755]: 测个 (2, 3) 的 stride。dilation 那个也是

File: oneflow/python/test/modules/test_conv.py [hunk 8]
```diff
@@ -0,0 +1,786 @@
...
+        _test_conv2d(
+            test_case,
+            conv,
+            test_conv2d_stride_data,
+            test_conv2d_stride_output,
+            test_conv2d_stride_weight,
+        )
+
+    def test_conv2d_kernel(test_case):
+        conv = flow.nn.Conv2d(1, 1, (5, 5), bias=False)
```
Comment (daquexian) [inline oneflow/python/test/modules/test_conv.py:765]: kernel size 也改成非正方形的

File: oneflow/python/test/modules/test_conv.py (hunk 6 above)
Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:745]: 好的

File: oneflow/python/test/modules/test_conv.py (hunk 7 above)
Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:755]: 好的

File: oneflow/python/test/modules/test_conv.py (hunk 8 above)
Comment (mosout) [reply oneflow/python/test/modules/test_conv.py:765]: 好的

**Review by daquexian — APPROVED** [review]
